### PR TITLE
Misc improvements

### PR DESCRIPTION
**Pull request by davidrapin**
Changes to `AdvancedMapping`:
- fixed bug that prevented node and relationship from being deleted from ES (https://github.com/graphaware/neo4j-to-elasticsearch/commit/9907b8ddd5fb9ea55a7ce91967f854b4bf63e52f)
- added a custom ES mapping to have a `not_analyzed` alternative of node label (`_labels.raw`) and relationship type (`_relationship.raw`) (https://github.com/graphaware/neo4j-to-elasticsearch/commit/6bf23553ca855acc8c8fc606173d1a0f64a670cf)

New Cypher procedures:
- added Cypher procedure (`ga.es.nodeMapping()` and `ga.es.relationshipMapping()`) to get current ES mapping (returns the same as `GET neo4j-index-node/node/_mapping` and `GET neo4j-index-relationship/relationship/_mapping`) (https://github.com/graphaware/neo4j-to-elasticsearch/commit/66aced00bd9fcaa622cd2a9ebee6b838ad2d0e9a)

``` JSON
{
   "neo4j-index-node": {
      "mappings": {
         "node": {
            "properties": {
               "_labels": {
                  "type": "string",
                  "fields": {
                     "raw": {
                        "type": "string",
                        "index": "not_analyzed"
                     }
                  }
               }
            }
         }
      }
   }
}
```
- added Cypher procedure (`ga.es.info()`) to get current ElasticSearch info (return the same as `GET /` on an ES server) (https://github.com/graphaware/neo4j-to-elasticsearch/commit/b639e6d9e42ee90d8e1f36fb75aa0e89ffe5ef7a)

``` JSON
{
   "name": "Fixer",
   "cluster_name": "elasticsearch",
   "version": {
      "number": "2.4.0",
      "build_hash": "ce9f0c7394dee074091dd1bc4e9469251181fc55",
      "build_timestamp": "2016-08-29T09:14:17Z",
      "build_snapshot": false,
      "lucene_version": "5.5.2"
   },
   "tagline": "You Know, for Search"
}
```

General improvement:
- improved general search performances by caching the `Searcher` instance in a ThreadLocal in `ElasticSearchProcedures` (https://github.com/graphaware/neo4j-to-elasticsearch/commit/8b1b0fa10b475892b25f3abfc22faaf2cbdde307)
- created `Key2IdResolver` to resolve UUID to Neo4j ID internally. Added a special case to allow for `KeyProperty=ID()` to use native Neo4j IDs as ElasticSearch IDs (instead of UUIDs). (https://github.com/graphaware/neo4j-to-elasticsearch/commit/2bb4e80033e97c020419f8efb4143e1ee67ecf5a)
- fixed intermittent exception `IllegalStateException: Connection pool shut down` (https://github.com/graphaware/neo4j-to-elasticsearch/commit/52e42d8a67a1b74948baaeaac868aa021db71055)

### CODE REVIEW

**Comment by davidrapin** [comment]
@ikwattro @bachmanm let me know what you think guys :)

**Comment by bachmanm** [comment]
will do after GC :)

**Comment by davidrapin** [comment]
Ok :p

File: src/main/java/com/graphaware/module/es/search/Key2IdResolver.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+package com.graphaware.module.es.search;
+
+import org.neo4j.graphdb.GraphDatabaseService;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static com.graphaware.runtime.RuntimeRegistry.getStartedRuntime;
+
+public class Key2IdResolver {
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/search/Key2IdResolver.java]: Please rename to `KeyToIdResolver`

File: src/main/java/com/graphaware/module/es/search/Key2IdResolver.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    private final Object uuidReader;
+    private final Method nodeUuidToId;
+    private final Method relUuidToId;
+
+    public Key2IdResolver(final GraphDatabaseService database, final String keyProperty) {
+
+        if (keyProperty.equals("ID()")) {
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/search/Key2IdResolver.java]: make `"ID()"` as `private static final String`

File: src/main/java/com/graphaware/module/es/util/JestClientFactory2.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.graphaware.module.es.util;
+
+import io.searchbox.client.JestClientFactory;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+public class JestClientFactory2 extends JestClientFactory {
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/util/JestClientFactory2.java]: Please find a better class name for this, maybe `CustomJestClientFactory` ?

**Comment by ikwattro** [comment]
@davidrapin KeyToIdResolver will need to be rebased with future changes, so I would wait a bit before merging the PR.

Changes will be related with #47

**Comment by ikwattro** [comment]
Also tests and documentation are missing

File: src/main/java/com/graphaware/module/es/search/SearchMatch.java [hunk 4]
```diff
@@ -19,12 +19,12 @@
...
 
 public class SearchMatch<T extends PropertyContainer> {
-    public final String uuid;
+    public final String key;
     public final Double score;
     private T item;
 
-    public SearchMatch(String uuid, double score) {
-        this.uuid = uuid;
+    public SearchMatch(String key, double score) {
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/search/SearchMatch.java]: this line needs rebase iaw #49

**Comment by davidrapin** [comment]
@ikwattro All requested changes done :)

**Comment by davidrapin** [comment]
@ikwattro ping :)

File: src/main/java/com/graphaware/module/es/proc/ElasticSearchProcedures.java [hunk 5]
```diff
@@ -86,9 +97,26 @@ private static Searcher getSearcher(GraphDatabaseService database) {
         return Stream.of(new JsonSearchResult(getSearcher(database).rawSearch(query, Relationship.class)));
     }
 
+    @Procedure("ga.es.nodeMapping")
+    @PerformsWrites
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/proc/ElasticSearchProcedures.java]: do those procedures really performs writes ?

File: src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+package com.graphaware.module.es.search.resolver;
+
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Result;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
```
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java]: this performance bottleneck should be documented

Comment (davidrapin) [inline src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java]: Right now this resolver is not usable.
It need to be tested and added to `ResolverFactory.createResolver` to be usable.
I kept this class for future implementation and testing only.

File: src/main/java/com/graphaware/module/es/proc/ElasticSearchProcedures.java (hunk 5 above)
Comment (davidrapin) [inline src/main/java/com/graphaware/module/es/proc/ElasticSearchProcedures.java]: good point, removing

File: src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java (hunk 6 above)
Comment (ikwattro) [inline src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java]: Allright, move this class to tests then

Comment (davidrapin) [inline src/main/java/com/graphaware/module/es/search/resolver/PropertyResolver.java]: I removed the class, I'll add it back when it's ready, in another PR :)

**Comment by davidrapin** [comment]
I just pushed the updated README, should be good now.

**Comment by davidrapin** [comment]
@ikwattro Sorry to bother you again but merging this would help me release the next version of Linkurious before Christmas :p 🎄 🎁

**Comment by ikwattro** [comment]
Great 👍  Sorry been very busy since arrived in italy :)

**Comment by ikwattro** [comment]
3.0.7.8 released

**Comment by ikwattro** [comment]
3.1.0.44.8 released as well, thanks a lot !

**Comment by davidrapin** [comment]
@ikwattro Thank you, congrats on the release :+1: Have a happy first Christmas in your new home in Italy :santa:

**Pull request merged**